### ISSUE TITLE
When typing or deleting text in a floating window, the cursor is also displayed in the top corner in the main window

### ISSUE THREAD

**Issue by YorickPeterse** [Lyude/neovim-gtk#29]
**Describe the bug**

This is a bit hard to describe, so here's a screencast:

https://user-images.githubusercontent.com/86065/166114002-76c79658-0565-411f-a8d1-cb41a466ab09.mp4

Note how as I type and later delete text in the Telescope window (but it applies to any floating window as far as I can tell), the cursor is sort of replicated in the top row of the main window.

**Technical information (please complete the following information):**
 - OS: Linux
 - Neovim version: 0.7.0
  - Neovim-Gtk build version: v0.4.1-6-gc36e2b5 (from source)

**Comment by Lyude** [comment]
FWIW - I marked this as a bug, but I'm going to remove that for the time being since I think this needs a little more investigation on my part as to who the culprit is here. If you know of any neovim UI clients that don't do this let me know, that'd probably help clarify.

**Comment by YorickPeterse** [comment]
@Lyude I think nvim-gtk is the _only_ one that _does_ do this, at least that I know of. Both the TUI and nvim-qt handle things as expected the last time I checked.

**Comment by Lyude** [comment]
> 

Good to know - I'll assume it's a bug then, at least until I get a chance to start digging into this (probably sometime after I get my ongoing GTK4 work done)

**Comment by Lyude** [comment]
Hah, good news! While I've been busy working on the gtk4 port of neovim-gtk, I discovered today I might actually know the reason this is happening and it's quite silly.
Basically - I never actually noticed the redraw handler in `src/nvim/redraw_handler.rs` interprets UI events totally incorrectly. We're supposed to take a set of UI events and process them such that we don't actually update the content the user sees until we receive a `flush`. It doesn't look at all like we're actually doing that though and instead we're just redrawing pretty much every time we receive any event, and simply ignoring flushes entirely.
Since I've had to change how we queue redraws generally in gtk4 due to the new snapshot renderer, this should be quite easy to fix for gtk4! I may or may not fix this in gtk3, but this would likely need a bit of additional work in order to use an offscreen cairo surface to maintain the intermediate screen buffer

**Comment by YorickPeterse** [comment]
@Lyude Thanks for the update. How usable is the GTK4 version of neovim-gtk? If it's in a usable state I'd be happy to try it for a while :smiley:

**Comment by Lyude** [comment]
> @Lyude Thanks for the update. How usable is the GTK4 version of neovim-gtk? If it's in a usable state I'd be happy to try it for a while smiley

Ah, sorry for the slow reply! It's certainly usable, but as I'm still finishing some small tidbits up with the conversion (been busy the last 2 months, so work has been slow but steady!) I haven't gotten to properly tweaking how rendering works to get it to optimize snapshot generation as much as possible, so it may be laggy in some places but seems to perform pretty alright in most of the scenarios I've tried it in. And some of the native widgets like the completion popover haven't been ported away from using `TreeView` widgets yet, which because of `TreeView` being on it's way out in GTK4 means they don't perform particularly well.

So - at this point I have most of the crashes fixed. Also, I've still got one cursor rendering regression to fix but that's very minor. Feel free to try the branch if you like, but be warned there's a lot of force pushing atm. I'll be cleaning that up pretty soon though with how close I am to having something that functionally works 1:1 with the GTK3 version.

Also - I realized after checking some other rendering issues I'm still not totally certain if GTK4 rendering will actually fix this, but it's certainly worth a try :).
